src/Compiler/Microsoft.AspNetCore.Razor.Language/src/DefaultAllowedChildTagDescriptorBuilder.cs

```diff
@@ -1,82 +1,67 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
-using System;
 using System.Collections.Generic;
-using System.Linq;
+using Microsoft.AspNetCore.Razor.PooledObjects;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-internal class DefaultAllowedChildTagDescriptorBuilder : AllowedChildTagDescriptorBuilder
+internal class DefaultAllowedChildTagDescriptorBuilder : AllowedChildTagDescriptorBuilder, IBuilder<AllowedChildTagDescriptor>
 {
     private readonly DefaultTagHelperDescriptorBuilder _parent;
-    private RazorDiagnosticCollection _diagnostics;
+    private RazorDiagnosticCollection? _diagnostics;
 
     public DefaultAllowedChildTagDescriptorBuilder(DefaultTagHelperDescriptorBuilder parent)
     {
         _parent = parent;
     }
 
-    public override string Name { get; set; }
+    public override string? Name { get; set; }
 
-    public override string DisplayName { get; set; }
+    public override string? DisplayName { get; set; }
 
-    public override RazorDiagnosticCollection Diagnostics
-    {
-        get
-        {
-            if (_diagnostics == null)
-            {
-                _diagnostics = new RazorDiagnosticCollection();
-            }
-
-            return _diagnostics;
-        }
-    }
+    public override RazorDiagnosticCollection Diagnostics => _diagnostics ??= new RazorDiagnosticCollection();
 
     public AllowedChildTagDescriptor Build()
     {
-        var diagnostics = Validate();
-        if (_diagnostics != null)
+        using var _ = HashSetPool<RazorDiagnostic>.GetPooledObject(out var diagnostics);
+
+        Validate(diagnostics);
+
+        if (_diagnostics is { } existingDiagnostics)
         {
-            diagnostics ??= new();
-            diagnostics.UnionWith(_diagnostics);
+            diagnostics.UnionWith(existingDiagnostics);
         }
 
         var displayName = DisplayName ?? Name;
+
         var descriptor = new DefaultAllowedChildTagDescriptor(
             Name,
             displayName,
-            diagnostics?.ToArray() ?? Array.Empty<RazorDiagnostic>());
+            diagnostics.ToArray());
 
         return descriptor;
     }
 
-    private HashSet<RazorDiagnostic> Validate()
+    private void Validate(HashSet<RazorDiagnostic> diagnostics)
     {
-        HashSet<RazorDiagnostic> diagnostics = null;
         if (string.IsNullOrWhiteSpace(Name))
         {
             var diagnostic = RazorDiagnosticFactory.CreateTagHelper_InvalidRestrictedChildNullOrWhitespace(_parent.GetDisplayName());
 
-            diagnostics ??= new();
             diagnostics.Add(diagnostic);
         }
         else if (Name != TagHelperMatchingConventions.ElementCatchAllName)
         {
-            foreach (var character in Name)
+            foreach (var character in Name!)
```

Comment:
probably a separate change, but seems odd that we can guarantee a name here

Reply:
It's because of the `string.IsNullOrWhitespace(...)` in the previous branch that we know that `Name` isn't null.

Reply:
I'm adding `IsNullOrWhitespace` and `IsNullOrEmpty` extension methods on string that are annotated for nullability. That should make these situations clearer.